File: hands-on/mock-signal.c
#include <signal.h>
#include <stdio.h>
#include <signal.h>
#include <sys/signal.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

void signal_handler(int signum) {
    printf("[received signal] %s\n", strsignal(signum));

    if (signum == SIGQUIT || signum == SIGINT) {
        printf("Terminating...\n");
        exit(0);
    }
}

int main() {
    // Ctrl+\ or Ctrl+C
    signal(SIGQUIT, signal_handler);
    signal(SIGINT, signal_handler);

    // kill -USR1 <pid>
    // kill -USR2 <pid>
    signal(SIGUSR1, signal_handler);
    signal(SIGUSR2, signal_handler);

    while(1) {
        printf("Press Ctrl+\\ or Ctrl+C to exit...(pid: %d)\n", getpid());
        sleep(2);
    }

    return 0;
}

File: hands-on/mock-signal.py
import signal
import time
import sys
import os

def handler(signum, frame):
    print("\n收到信号，准备退出...")

    time.sleep(1.5)
    print("清理完成，再见！")

    sys.exit(0)

# 注册信号处理器 (Ctrl+C, Ctrl+\)
signal.signal(signal.SIGINT, handler)
signal.signal(signal.SIGQUIT, handler)


print(f"开始工作，按 Ctrl+C 或 Ctrl + \\ 退出 (pid: {os.getpid()})...")

while True:
    time.sleep(1)  # 模拟工作
